### DISCUSSION TITLE
Reference SG generated class in XAML

### DISCUSSION THREAD

**Discussion by rabbitism** [AvaloniaUI/Avalonia#16081] [Q&A]
Hello devs,
  I have a class generated by source generator and I need to reference it in xaml. It works perfectly fine before building, but it failed building with error "class doesn't exist". Seems like XAML compilation happens before SG, is there any work around for this?

**Comment by aligungr** [comment]
If you are using Roslyn's (relatively) new source generator mechanism, then I think it's currently not possible.

Avalonia also uses source generators internally for XAML compilation, and generators cannot see each others' sources, and it's by design.

See:
https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#proposal

From the link:
_"Run un-ordered, each generator will see the **same input compilation**, with no access to files created by other source generators."_

**Reply by maxkatz6** [reply]
> Avalonia also uses source generators internally for XAML compilation, and generators cannot see each others' sources, and it's by design.

That's partially true. Avalonia x:Name generator is a source generator. But XAML compilation itself isn't, and happens after c# compilation.

**Comment by maxkatz6** [comment]
XAML compilation happens after C# compilation.

But avalonia property generator happens in order with other source generators, so it might produce undefined behavior not controlled by avalonia.

**Comment by workgroupengineering** [comment]
The workaround I use is to put the entire generated class in a library then add a reference to it in the project where xaml is located.
